packages/patrol_cli/lib/src/coverage/run_code_coverage.dart

```diff
@@ -0,0 +1,190 @@
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:coverage/coverage.dart';
+import 'package:glob/glob.dart';
+import 'package:patrol_cli/src/base/logger.dart';
+import 'package:patrol_cli/src/devices.dart';
+import 'package:vm_service/vm_service.dart';
+import 'package:vm_service/vm_service_io.dart';
+
+Future<Map<String, HitMap>> _collectCoverage(
+  VmService client,
+  Uri vmUri,
+  String packageName,
+  String mainIsolateId,
+) async {
+  final coverage = await collect(
+    vmUri,
+    false,
+    false,
+    false,
+    {packageName},
+  );
+
+  final socket = await WebSocket.connect(client.wsUri!)
+    ..add(
+      jsonEncode(
+        {
+          'jsonrpc': '2.0',
+          'id': 21,
+          'method': 'ext.patrol.markTestCompleted',
+          'params': {
+            'isolateId': mainIsolateId,
+            'command': 'markTestCompleted',
+          },
+        },
+      ),
+    );
+  await socket.close();
+
+  final map = await HitMap.parseJson(
+    coverage['coverage'] as List<Map<String, dynamic>>,
+  );
+
+  return map;
+}
+
+Future<ProcessResult> _forwardAdbPort(String host, String guest) async {
+  return Process.run('adb', ['forward', 'tcp:$host', 'tcp:$guest']);
```

Comment:
IIRC this will blow up on Windows, always pass `runInShell: true` in such cases

example:

https://github.com/leancodepl/patrol/blob/9f97af3615374ba8d5996e5ba6ad75496d72b7c5/packages/adb/lib/src/adb.dart#L199

Reply:
Also, don't use `Process` from `dart:io`, use `package:process` to allow for dependency injection (required for testing). CTRL+F "package:process" to see examples in this repo

Reply:
Also, if you look around patrol, you'll notice that a wrapper around [adb](https://pub.dev/packages/adb) is used.